File: leetcode/350. Intersection of Two Arrays II.cpp
// Copyright 2016 Qi Wang
// Date: 2016-10-06
class Solution {
 public:
  vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
    auto cnts1 = count(nums1);
    auto cnts2 = count(nums2);
    vector<int> res;
    for (auto cnt : cnts1) {
      int times = min(cnt.second, cnts2[cnt.first]);
      for (int i = 0; i < times; ++i) res.push_back(cnt.first);
    }
    return res;
  }

 private:
  inline unordered_map<int, int> count(const vector<int>& nums) const {
    unordered_map<int, int> cnts;
    for (int num : nums) ++cnts[num];
    return cnts;
  }
};

File: leetcode/350. Intersection of Two Arrays II.py
class Solution:
  def intersect(self, nums1, nums2):
    """
    :type nums1: List[int]
    :type nums2: List[int]
    :rtype: List[int]
    """
    # if len(nums1) > len(nums2):
    #   return self.intersect(nums2, nums1)
    # Constructs a map
    num_to_freq_map = {}
    for num in nums1:
      if num not in num_to_freq_map:
        num_to_freq_map.setdefault(num, 1)
      else:
        num_to_freq_map[num] = num_to_freq_map[num] + 1
    print(num_to_freq_map)
    # Loops nums2 to get the result by looking up the map
    result = []
    for num in nums2:
      if num in num_to_freq_map and num_to_freq_map[num] > 0:
        num_to_freq_map[num] = num_to_freq_map[num] - 1
        result.append(num)
    return result
